### ISSUE TITLE
Rustc Confuses Associated Types on Type Inference

### ISSUE THREAD

**Issue by apopiak** [rust-lang/rust#82513]
The following code does not compile ([playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4ca706e8932e3fb5f4309389473554ee)):

```rust
pub trait Trait {
	type Value;
	type AccountId: Default;
}

pub enum A<T: Trait> {
	_C(
		(T::AccountId, T::AccountId),
	),
}

impl<T: Trait> A<T>
where
	(T::Value, T::Value): Default,
{
	fn foo() -> Self {
		A::_C(Default::default())
	}
}
```

I expected to see this happen: Successful compilation

Instead, this happened: I get `E0271`
```
error[E0271]: type mismatch resolving `<T as Trait>::AccountId == <T as Trait>::Value`
  --> src/lib.rs:17:3
   |
17 |         A::_C(Default::default())
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^ expected Trait::AccountId, found Trait::Value
   |
   = note: expected associated type `<T as Trait>::AccountId`
              found associated type `<T as Trait>::Value`
```
Ways to get this to compile (that shouldn't matter/don't change the semantics):
+ remove the `where` bound
+ use explicit call to default `<(T::AccountId, T::AccountId)>::default()` 

### Meta
`rustc --version --verbose`:
```
rustc 1.50.0 (cb75ad5db 2021-02-10)
binary: rustc
commit-hash: cb75ad5db02783e8b0222fee363c5f63f7e2cf5b
commit-date: 2021-02-10
host: x86_64-unknown-linux-gnu
release: 1.50.0
```
(also happens on 1.52-nightly)

**Comment by b-naber** [comment]
Didn't you mean to write?

```
impl<T: Trait> A<T>
where
	(T::AccountId, T::AccountId): Default,
{
	fn foo() -> Self {
		A::_C(Default::default())
	}
}
```

That seems more like what was intended here and it compiles.

**Comment by kadiwa4** [comment]
This appears to be fixed-by-next-solver, but I don't think I can add that label.

**Comment by apopiak** [comment]
> Didn't you mean to write?
> 
> ```
> impl<T: Trait> A<T>
> where
> 	(T::AccountId, T::AccountId): Default,
> {
> 	fn foo() -> Self {
> 		A::_C(Default::default())
> 	}
> }
> ```

Nope, the bug is specifically about the compiler getting confused by the useless/superfluous (at least with this particular simplified code example) where clause that should not have any impact on the compilation.
It is a reduced version of real life code IIRC (it's been a while 😅 ) that might be generated.
